contracts/utils/RetireToucanCarbon.sol

```diff
@@ -0,0 +1,1405 @@
+            abi.encodeWithSelector(
+                token.approve.selector,
+                spender,
+                newAllowance
+            )
+        );
+    }
+
+    function safeDecreaseAllowance(
+        IERC20 token,
+        address spender,
+        uint256 value
+    ) internal {
+        unchecked {
+            uint256 oldAllowance = token.allowance(address(this), spender);
+            require(
+                oldAllowance >= value,
+                "SafeERC20: decreased allowance below zero"
+            );
+            uint256 newAllowance = oldAllowance - value;
+            _callOptionalReturn(
+                token,
+                abi.encodeWithSelector(
+                    token.approve.selector,
+                    spender,
+                    newAllowance
+                )
+            );
+        }
+    }
+
+    /**
+     * @dev Imitates a Solidity high-level call (i.e. a regular function call to a contract), relaxing the requirement
+     * on the return value: the return value is optional (but if data is returned, it must not be false).
+     * @param token The token targeted by the call.
+     * @param data The call data (encoded using abi.encode or one of its variants).
+     */
+    function _callOptionalReturn(IERC20 token, bytes memory data) private {
+        // We need to perform a low level call here, to bypass Solidity's return data size checking mechanism, since
+        // we're implementing it ourselves. We use {Address.functionCall} to perform this call, which verifies that
+        // the target address contains contract code and also asserts for success in the low-level call.
+
+        bytes memory returndata = address(token).functionCall(
+            data,
+            "SafeERC20: low-level call failed"
+        );
+        if (returndata.length > 0) {
+            // Return data is optional
+            require(
+                abi.decode(returndata, (bool)),
+                "SafeERC20: ERC20 operation did not succeed"
+            );
+        }
+    }
+}
+
+// File: contracts/SafeERC20.sol
+
+// File: contracts/RetireToucanCarbon.sol
+
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+contract RetireToucanCarbon is Ownable {
+    using SafeERC20 for IERC20;
+
+    constructor(
+        address _KLIMA,
+        address _sKLIMA,
+        address _wsKLIMA,
+        address _USDC,
+        address _staking,
+        address _stakingHelper,
+        address _DAO
+    ) {
+        KLIMA = _KLIMA;
+        sKLIMA = _sKLIMA;
+        wsKLIMA = _wsKLIMA;
+        USDC = _USDC;
+        staking = _staking;
+        stakingHelper = _stakingHelper;
+        DAO = _DAO;
+    }
+
+    address public immutable KLIMA;
+    address public immutable sKLIMA;
+    address public immutable wsKLIMA;
+    address public immutable USDC;
+    address public immutable staking;
+    address public immutable stakingHelper;
+    address public immutable DAO;
+
+    event ToucanRetired(
```

Comment:
Could be indexed:
```
event ToucanRetired(
        address indexed retiree,
        string indexed beneficiary,
        address indexed carbonPool,
        address carbonToken,
        uint256 amount
    );
```

Reply:
Indexed a few of the addresses, as the total fields in this event have now expanded due to feedback.

Reply:
Spotted it!
For my part, I prefer to log always `msg.sender` that lets me follow value flow easily.

Reply:
Please mind that according to the latest specification we will implement the following attributes:
```
    struct Data {
        uint256 projectVintageTokenId;
        uint256 retiredCO2inKg;   // @dev amount in TCO2 gets divided by 1e15
        address retiringEntity;   /// @dev msg.sender in TCO2 contract calling the function
        address beneficiary;
        string beneficiaryString;/
        string retirementMessage;
        string tokenURI; /// @dev not being set on mint
    }